aspnetcore/test/middleware.md

```diff
@@ -0,0 +1,79 @@
+---
+title: Test ASP.NET Core middleware
+author: cross
+description: Learn how to test ASP.NET Core middleware with TestServer.
+ms.author: riande
+ms.custom: mvc
+ms.date: 4/05/2019
+no-loc: [Blazor, "Identity", "Let's Encrypt", Razor, SignalR]
+uid: test/middleware
+---
+# Test ASP.NET Core middleware
+
+By [Chris Ross](https://github.com/Tratcher)
+
+Middleware can be tested in isolation using the <xref:Microsoft.AspNetCore.TestHost.TestServer>. Instantiate an app pipeline containing only the components that you need to test and send custom requests to verify middleware behavior.
+
+The requests:
+
+* Are sent in-memory rather than being serialized over the network.
+* Avoid additional concerns like port management and HTTPS certificates.
+
+Exceptions in the middleware can flow directly back to the calling test.
```

Comment:
```suggestion
Advantages:

* Requests are sent in-memory rather than being serialized over the network.
* This avoids additional concerns like port management and HTTPS certificates.
* Exceptions in the middleware can flow directly back to the calling test.
* It's possible to customize server data structures like HttpContext directly in the test.
```